rafthttp/stream.go

```diff
@@ -0,0 +1,279 @@
+		if w != nil {
+			w.Close()
+		}
+		if nw != nil {
+			heartbeatc, msgc = tickc, cw.msgc
+		} else {
+			heartbeatc, msgc = nil, nil
+		}
+		w = nw
+		cw.mu.Lock()
+		cw.working = nw != nil
+		cw.mu.Unlock()
+	}
+	for {
+		select {
+		case <-heartbeatc:
+			// send empty message as heartbeat
+			if err := binary.Write(w, binary.BigEndian, uint64(0)); err != nil {
+				log.Printf("rafthttp: encountered error heartbeating to connection: %v", err)
+				reset(nil)
+				continue
+			}
+			w.Flush()
+		case m := <-msgc:
+			if !cw.enc.isEllegible(m) {
+				log.Printf("rafthttp: silently dropped message %s because it is not ellegible", m.Type)
+				continue
+			}
+			if err := cw.enc.encodeTo(w, m); err != nil {
+				log.Printf("rafthttp: encountered error writing to connection: %v", err)
+				reset(nil)
+				continue
+			}
+			w.Flush()
+		case conn := <-cw.connc:
+			cw.enc.loadMetadata(conn.info)
+			reset(conn.WriteFlushCloser)
+		case <-cw.stopc:
+			if w != nil {
+				w.Close()
+			}
+			close(cw.done)
+			return
+		}
+	}
+}
+
+func (cw *streamWriter) canSend(m raftpb.Message) bool {
+	if !cw.enc.isEllegible(m) {
+		return false
+	}
+	cw.mu.Lock()
+	defer cw.mu.Unlock()
+	return cw.working
+}
+
+func (cw *streamWriter) setOutgoingConn(conn outgoingConn) bool {
+	select {
+	case cw.connc <- conn:
+		return true
+	case <-cw.done:
+		return false
+	}
+}
+
+func (cw *streamWriter) stop() {
+	close(cw.stopc)
+	<-cw.done
+}
+
+type streamDialer struct {
+	mu      sync.RWMutex // guard field u
+	u       string
+	tr      http.RoundTripper
+	from    types.ID
+	to      types.ID
+	cid     types.ID
+	typeStr string
+}
+
+func newStreamDialer(u string, tr http.RoundTripper, from, to, cid types.ID, typeStr string) *streamDialer {
+	return &streamDialer{
+		u:       u,
+		tr:      tr,
+		from:    from,
+		to:      to,
+		cid:     cid,
+		typeStr: typeStr,
+	}
+}
+
+func (d *streamDialer) dial(dec decoder) (io.ReadCloser, error) {
+	d.mu.RLock()
+	uu, err := url.Parse(d.u)
+	d.mu.RUnlock()
+	if err != nil {
+		return nil, fmt.Errorf("parse url %s error: %v", d.u, err)
+	}
+	uu.Path = path.Join(RaftStreamPrefix, d.from.String())
+	req, err := http.NewRequest("GET", uu.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("new request to %s error: %v", d.u, err)
+	}
+	req.Header.Set("X-Etcd-Cluster-ID", d.cid.String())
+	req.Header.Set("X-Raft-To", d.to.String())
+	req.Header.Set("X-Stream-Type", d.typeStr)
+	dec.saveMetadata(req.Header)
+	resp, err := d.tr.RoundTrip(req)
+	if err != nil {
+		return nil, fmt.Errorf("error posting to %q: %v", d.u, err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		resp.Body.Close()
+		return nil, fmt.Errorf("unhandled http status %d", resp.StatusCode)
+	}
+	return resp.Body, nil
+}
+
+func (d *streamDialer) update(u string) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.u = u
+}
+
+type recvFunc func(m raftpb.Message) error
+
+type streamReader struct {
+	dialer *streamDialer
+	dec    decoder
+	recv   recvFunc
```

Comment:
recvFunc uses channel inside to do communication.
I use recvFunc here because the same logic is also used at https://github.com/coreos/etcd/pull/2323/files#diff-288ade94a6578e3a03cc7336398b3565R98
I understand a channel makes it more golang style, but it does harm by increasing the duplicated code here.